keps/sig-scheduling/1923-try-nominated-node-first/README.md

```diff
@@ -0,0 +1,220 @@
+new scheduling cycle before looping all over the nodes in the cluster for the
+phase of the filtering.
+
+## Motivation
+
+In real production environment, pods can have different priorites due to business
+needs, the preemption could happen to make sure high priority pods could got scheduled
+and spawned.
+
+For the cluster with huge number of computing nodes, looping over all the nodes in
+the cluster for the phase of the filtering is an unnegligible consuming for scheduler.
+
+### Goals
+
+- In the case of `pod.Status.NominatedNodeName` is set for the incoming pod, scheduler
+  will try the nominated node first for the filtering phase, failed to schedule on the
+  nominated node will fall back to verify all the nodes in the cluster instead of preemption
+  again.
+
+## Proposal
+
+### User Stories (Optional)
+
+As a premium user, I want my high priority application got spin up more quickly,
+I don't care where the application is running, whether there is any other nodes
+in the cluster have more disk space, memory size etc., as long as my application
+run normally.
+
+### Notes/Constraints/Caveats (Optional)
+
+When this feature is enabled the preemptor Pod might not be dispatched to the best
+candidated node in some corner case, e.g. another node releases the resources and
+becomes the best candidate while the victim pods got removed from the nominated node.
+
+## Design Details
+
+This feature hopes to improve the performance with the minimal change on the existing scheduling
+framework, scheduler will short-circuit to check the nominated node directly instead of looping
+all the nodes in the cluster, and if the nominated node doesn't fit, a fresh new scheduling cycle
+will be started to get an new node for scheduling.
+
+In most situation, the new scheduling cycle is not needed, but we still got the case like the
+topology changing, victim pods deletion is still on-the-fly in the following scheduling cycle.
+Those cases will make the nominated node not fits for the coming pod anymore, and a new scheduling
+cycle is needed in those case. This is no harm since it will eventually get convergent.
+
+### Implementation Details
+
+1. In filtering phase, which is currently implemented in the method of `findNodesThatFitPod`,
+   check the nominated directly if the incoming pod has the `pod.Status.NominatedNodeName` defined
+   and the feature gate is enabled.
+2. In case the nominated node not suit the incoming pod anymore, return `ErrNominateNode` instead of
+   `core.FitError`, because this will give scheduler a chance to clean up the nominated node from
+   the pod and find a new node to schedule instead of preemption again (might already have during
+   the period). i.e. a fresh new scheduling cycle will be started from here.
+   A new error `ErrNominateNode` should be defined to describe what's going wrong on the nominated node.
+3. If the nominated node is not suit for the pod anymore, the scheduling failure will be recorded
```

Comment:
The enumeration makes it feel like 3 happens after 2, but in reality 3 is giving more details on 2. I don't think we need the paragraph at all.

Reply:
Done